server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricDao.java

```diff
@@ -0,0 +1,156 @@
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.ibatis.session.RowBounds;
+import org.sonar.db.Dao;
+import org.sonar.db.DbSession;
+import org.sonar.db.RowNotFoundException;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static org.sonar.db.DatabaseUtils.executeLargeInputs;
+import static org.sonar.db.DatabaseUtils.executeLargeInputsWithoutOutput;
+
+public class MetricDao implements Dao {
+
+  @CheckForNull
+  public MetricDto selectByKey(DbSession session, String key) {
+    return mapper(session).selectByKey(key);
+  }
+
+  public List<MetricDto> selectByKeys(final DbSession session, List<String> keys) {
+    return executeLargeInputs(keys, mapper(session)::selectByKeys);
+  }
+
+  public MetricDto selectOrFailByKey(DbSession session, String key) {
+    MetricDto metric = selectByKey(session, key);
+    if (metric == null) {
+      throw new RowNotFoundException(String.format("Metric key '%s' not found", key));
+    }
+    return metric;
+  }
+
+  public List<MetricDto> selectAll(DbSession session) {
+    return mapper(session).selectAll();
+  }
+
+  public List<MetricDto> selectEnabled(DbSession session) {
+    return mapper(session).selectAllEnabled();
+  }
+
+  public List<MetricDto> selectEnabled(DbSession session, @Nullable Boolean isCustom, int offset, int limit) {
+    Map<String, Object> properties = Maps.newHashMapWithExpectedSize(1);
+    if (isCustom != null) {
+      properties.put("isCustom", isCustom);
+    }
+
+    return mapper(session).selectAllEnabled(properties, new RowBounds(offset, limit));
+  }
+
+  public int countEnabled(DbSession session, @Nullable Boolean isCustom) {
+    return mapper(session).countEnabled(isCustom);
+  }
+
+  public MetricDto insert(DbSession session, MetricDto dto) {
+    mapper(session).insert(dto);
+
+    return dto;
+  }
+
+  public void insert(DbSession session, Collection<MetricDto> items) {
+    for (MetricDto item : items) {
+      insert(session, item);
+    }
+  }
+
+  public void insert(DbSession session, MetricDto item, MetricDto... others) {
+    insert(session, Lists.asList(item, others));
+  }
+
+  public List<String> selectEnabledDomains(DbSession session) {
+    return newArrayList(Collections2.filter(mapper(session).selectDomains(), new NotEmptyPredicate()));
+  }
+
+  public List<MetricDto> selectAvailableCustomMetricsByComponentUuid(DbSession session, String projectUuid) {
+    return mapper(session).selectAvailableCustomMetricsByComponentUuid(projectUuid);
+  }
+
+  public List<MetricDto> selectByIds(DbSession session, Set<Integer> idsSet) {
+    return executeLargeInputs(new ArrayList<>(idsSet), mapper(session)::selectByIds);
+  }
+
+  private static class NotEmptyPredicate implements Predicate<String> {
+
+    @Override
+    public boolean apply(@Nonnull String input) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2638)